### PR TITLE
Impl service errors

### PR DESCRIPTION
**Pull request by tgiardina**
This is in response to #21.

1. I added an `enum` called `ServiceError` in `constants`.
2. I replaced error codes in `controllers/user` and `services/user` with `ServiceError`s.

This both prevents (1) silly bugs arising from error code typos, and (2) the creation of redundant error codes. But it also adds an additional abstraction that a new user will have to decipher. I think the trade off is worth it.

@kunwar97 You said you were interested in working on this, so let me know if you think there's a better way of doing it.

### CODE REVIEW

**Comment by kunwar97** [comment]
@tgiardina Instead of creating enums we can create classes for errors which will contain all information regarding that error.

First we create a base class HttpError which contains message, statusCode, errors, meta(for extra info).
 
If we want throw UserNotFound Error then  we create another class user-not-found.error.ts which extends HttpError class and in constructor we pass error message message

**Comment by kunwar97** [comment]
```
export class HttpException extends Error {
message: string;
statusCode: number;
errors?: any;
meta?: any;

constructor(message: string, statusCode: number, errors?: any, meta?:any) {
    super(message);

    this.message    = message;
    this.statusCode = statusCode;
    this.meta   = meta;
    this.errors = errors;
  }

}
```

```
export class ModelNotFoundException extends HttpException {
    constructor(message: string, errors?: any, meta?: any) {
        super(message, errorCode, 404, errors);
    }
}
```

**Comment by tgiardina** [comment]
@kunwar97 I'll start off by saying that if we were working on an actual project for a client, your approach would absolutely be the way to go! (I would probably exclude the status code because I think only the controller layer should worry about status, but I would want to use the new error classes in our domain layer (i.e., `services`, `repositories`, `models`) as well.)

Where this gets tricky is deciding how to balance [Goals 2 and 3](https://github.com/tgiardina/typescript-typeorm-realworld-example-app#goals). Creating error classes definitely adheres to Goal 2, but it break Goal 3. In particular, this project should show off how to use `inversify` together with `TypeORM`, and adding error classes doesn't do anything to do that.

I added the `result` helper class as well as the `ServiceError` `enum` in hopes to find a balance between Goal 2 and 3 -- adding some structure without going so far as creating error classes. But maybe that's not the way to go. Looking at other Real World backends (e.g., [node-express](https://github.com/gothinkster/node-express-realworld-example-app) and [nestJS](https://github.com/lujakob/nestjs-realworld-example-app)), they handle their errors naively (using promises and try-catch blocks). I think these projects are pretty good, so I'm inclined to follow their lead.

In summary,

- I think your way is best for actual projects.
- I think naively handling errors is probably best for this demo (based on what  [node-express](https://github.com/gothinkster/node-express-realworld-example-app) and [nestJS](https://github.com/lujakob/nestjs-realworld-example-app) have done).
- I think the `result` helper and `ServiceError` `enum` that I implemented fall in between the two and don't accomplish much.

So unless you disagree, I think I will delete the `result` helper and close this pull request without merging it.
